Add wrappers for Nevergrad

Wrap the gradient free optimizers from [nevergrad](https://github.com/facebookresearch/nevergrad) in optimagic.

Nevergrad implements the following algorithms:

- `NgIohTuned` is "meta"-optimizer which adapts to the provided settings (budget, number of workers, parametrization) and should therefore be a good default.
- `TwoPointsDE` is excellent in many cases, including very high `num_workers`.
- `PortfolioDiscreteOnePlusOne` is excellent in discrete settings of mixed settings when high precision on parameters is not relevant; it's possibly a good choice for hyperparameter choice.
- `OnePlusOne` is a simple robust method for continuous parameters with `num_workers` < 8.
- `CMA` is excellent for control (e.g. neurocontrol) when the environment is not very noisy (num_workers ~50 ok) and when the budget is large (e.g. 1000 x the dimension).
- `TBPSA` is excellent for problems corrupted by noise, in particular overparameterized (neural) ones; very high `num_workers` ok).
- `PSO` is excellent in terms of robustness, high `num_workers` ok.
- `ScrHammersleySearchPlusMiddlePoint` is excellent for super parallel cases (fully one-shot, i.e. `num_workers` = budget included) or for very multimodal cases (such as some of our MLDA problems); don't use softmax with this optimizer.
- `RandomSearch` is the classical random search baseline; don't use softmax with this optimizer.

In the long run we want to wrap all of them, **but if you are tackling this as your first issue you should focus on one. In that case please coment below which optimizer you are going to work on so we don't duplicate efforts.**

I have written some basic structure of code and tested locally by installing nevergrad in optimagic conda environment.
But pre-commit is giving this error. I have added nevergrad to environments.yml too.
src/optimagic/optimizers/nevergrad_optimizers.py:7: error: Cannot find implementation or library stub for module named "nevergrad"  [import-not-found]

I am using ubuntu 24.04.

[https://github.com/PyCQA/docformatter] 
==> File /tmp/tmpivrx7hl6/.pre-commit-hooks.yaml
==> At Hook(id='docformatter-venv')
==> At key: language
=====> Expected one of conda, coursier, dart, docker, docker_image, dotnet, fail, golang, haskell, julia, lua, node, perl, pygrep, python, r, ruby, rust, script, swift, system but got: 'python_venv'

--- a/.tools/envs/testenv-linux.yml
+++ b/.tools/envs/testenv-linux.yml
@@ -8,6 +8,7 @@ dependencies:
   - jax
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
+  - nevergrad  # dev, tests
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
   - pytest  # dev, tests

--- a/.tools/envs/testenv-linux.yml
+++ b/.tools/envs/testenv-linux.yml
@@ -8,7 +8,6 @@ dependencies:
   - jax
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
-  - nevergrad  # dev, tests
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
   - pytest  # dev, tests
@@ -30,6 +29,7 @@ dependencies:
   - annotated-types  # dev, tests
   - iminuit  # dev, tests
   - pip:  # dev, tests, docs
+      - nevergrad  # dev, tests
       - DFO-LS>=1.5.3  # dev, tests
       - Py-BOBYQA  # dev, tests
       - fides==0.7.4  # dev, tests

--- a/.tools/envs/testenv-numpy.yml
+++ b/.tools/envs/testenv-numpy.yml
@@ -8,6 +8,7 @@ dependencies:
   - numpy<2
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
+  - nevergrad  # dev, tests
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
   - pytest  # dev, tests

--- a/.tools/envs/testenv-numpy.yml
+++ b/.tools/envs/testenv-numpy.yml
@@ -8,7 +8,6 @@ dependencies:
   - numpy<2
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
-  - nevergrad  # dev, tests
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
   - pytest  # dev, tests
@@ -28,6 +27,7 @@ dependencies:
   - annotated-types  # dev, tests
   - iminuit  # dev, tests
   - pip:  # dev, tests, docs
+      - nevergrad  # dev, tests
       - DFO-LS>=1.5.3  # dev, tests
       - Py-BOBYQA  # dev, tests
       - fides==0.7.4  # dev, tests

--- a/.tools/envs/testenv-others.yml
+++ b/.tools/envs/testenv-others.yml
@@ -6,7 +6,6 @@ channels:
 dependencies:
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
-  - nevergrad  # dev, tests
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
   - pytest  # dev, tests
@@ -28,6 +27,7 @@ dependencies:
   - annotated-types  # dev, tests
   - iminuit  # dev, tests
   - pip:  # dev, tests, docs
+      - nevergrad  # dev, tests
       - DFO-LS>=1.5.3  # dev, tests
       - Py-BOBYQA  # dev, tests
       - fides==0.7.4  # dev, tests

--- a/.tools/envs/testenv-others.yml
+++ b/.tools/envs/testenv-others.yml
@@ -6,6 +6,7 @@ channels:
 dependencies:
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
+  - nevergrad  # dev, tests
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
   - pytest  # dev, tests

--- a/.tools/envs/testenv-pandas.yml
+++ b/.tools/envs/testenv-pandas.yml
@@ -8,6 +8,7 @@ dependencies:
   - numpy<2
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
+  - nevergrad  # dev, tests
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
   - pytest  # dev, tests

--- a/.tools/envs/testenv-pandas.yml
+++ b/.tools/envs/testenv-pandas.yml
@@ -8,7 +8,6 @@ dependencies:
   - numpy<2
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
-  - nevergrad  # dev, tests
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
   - pytest  # dev, tests
@@ -28,6 +27,7 @@ dependencies:
   - annotated-types  # dev, tests
   - iminuit  # dev, tests
   - pip:  # dev, tests, docs
+      - nevergrad  # dev, tests
       - DFO-LS>=1.5.3  # dev, tests
       - Py-BOBYQA  # dev, tests
       - fides==0.7.4  # dev, tests

--- a/docs/source/algorithms.md
+++ b/docs/source/algorithms.md
@@ -3984,6 +3984,61 @@ iminuit).
       - Values greater than 1 specify the maximum number of restart attempts.  
 ```
 
+(nevergrad-algorithms)=
+
+## Nevergrad Optimizers
+
+optimagic supports some algorithms from the
+[Nevergrad](https://facebookresearch.github.io/nevergrad/index.html) library. To use
+these optimizers, you need to have
+[the nevergrad package](https://github.com/facebookresearch/nevergrad) installed.
+(`pip install nevergrad`).
+
+```{eval-rst}
+.. dropdown:: nevergrad_pso
+
+    .. code-block::
+
+        "nevergrad_pso"
+
+    Minimize a scalar function using the Particle Swarm Optimization (PSO) algorithm.
+
+    The Particle Swarm Optimization algorithm was originally proposed by
+    :cite:`Kennedy1995`. The implementation in Nevergrad is based on
+    :cite:`Zambrano2013`.
+
+    Particle Swarm Optimization (PSO) solves a problem by having a population of
+    candidate solutions, here dubbed particles, and moving these particles around in the
+    search-space according to simple mathematical formulae over the particle's position
+    and velocity. Each particle's movement is influenced by its local best known
+    position (termed "cognitive" component), but is also guided toward the best known
+    positions (termed "social" component) in the search-space, which are updated as
+    better positions are found by other particles. This is expected to move the swarm
+    toward the best solutions.
+
+    - **transform** (str): The transform to use to map from PSO optimization space to
+      R-space.
+    - **population_size** (int): Population size of the particle swarm.
+    - **n_cores** (int): Number of cores to use.
+    - **stopping.maxfun** (int): Maximum number of function evaluations.
+    - **inertia** (float): Inertia weight. Denoted by :math:`\omega`.
+      Default is 0.7213475204444817. To prevent divergence, the value must be smaller
+      than 1. It controls the influence of the particle's previous velocity on its
+      movement.
+    - **cognitive** (float): Cognitive coefficient. Denoted by :math:`\phi_p`.
+      Default is 1.1931471805599454. Typical values range from 1.0 to 3.0. It controls
+      the influence of its own best known position on the particle's movement.
+    - **social** (float): Social coefficient. Denoted by :math:`\phi_g`.
+      Default is 1.1931471805599454. Typical values range from 1.0 to 3.0. It controls
+      the influence of the swarm's best known position on the particle's movement.
+    - **quasi_opp_init** (bool): Whether to use quasi-opposition initialization.
+      Default is False.
+    - **speed_quasi_opp_init** (bool): Whether to use quasi-opposition initialization
+      for speed. Default is False.
+    - **special_speed_quasi_opp_init** (bool): Whether to use special quasi-opposition
+      initialization for speed. Default is False.
+```
+
 ## References
 
 ```{eval-rst}

--- a/docs/source/algorithms.md
+++ b/docs/source/algorithms.md
@@ -4017,9 +4017,13 @@ these optimizers, you need to have
     toward the best solutions.
 
     - **transform** (str): The transform to use to map from PSO optimization space to
-      R-space.
+      R-space. Available options are:
+      - "arctan" (default)
+      - "identity"
+      - "gaussian"
     - **population_size** (int): Population size of the particle swarm.
     - **n_cores** (int): Number of cores to use.
+    - **seed** (int): Seed used by the internal random number generator.
     - **stopping.maxfun** (int): Maximum number of function evaluations.
     - **inertia** (float): Inertia weight. Denoted by :math:`\omega`.
       Default is 0.7213475204444817. To prevent divergence, the value must be smaller

--- a/docs/source/refs.bib
+++ b/docs/source/refs.bib
@@ -906,4 +906,38 @@ url = {https://www.sciencedirect.com/science/article/pii/0010465575900399},
 author = {F. James and M. Roos}
 }
 
+@article{JAMES1975343,
+title = {Minuit - a system for function minimization and analysis of the parameter errors and correlations},
+journal = {Computer Physics Communications},
+volume = {10},
+number = {6},
+pages = {343-367},
+year = {1975},
+issn = {0010-4655},
+doi = {https://doi.org/10.1016/0010-4655(75)90039-9},
+url = {https://www.sciencedirect.com/science/article/pii/0010465575900399},
+author = {F. James and M. Roos}
+}
+
+@InProceedings{Kennedy1995,
+  author={Kennedy, J. and Eberhart, R.},
+  booktitle={Proceedings of ICNN'95 - International Conference on Neural Networks}, 
+  title={Particle swarm optimization}, 
+  year={1995},
+  volume={4},
+  pages={1942-1948 vol.4},
+  keywords={Particle swarm optimization;Birds;Educational institutions;Marine animals;Testing;Humans;Genetic algorithms;Optimization methods;Artificial neural networks;Performance evaluation},
+  doi={10.1109/ICNN.1995.488968},
+}
+
+@InProceedings{Zambrano2013,
+  author = {Zambrano-Bigiarini, Mauricio and Clerc, Maurice and Rojas, Rodrigo},
+  booktitle = {2013 IEEE Congress on Evolutionary Computation}, 
+  title = {Standard Particle Swarm Optimisation 2011 at CEC-2013: A baseline for future PSO improvements}, 
+  year = {2013},
+  pages = {2337-2344},
+  keywords = {Optimization;Standards;Benchmark testing;Topology;Algorithm design and analysis;Convergence;Equations;particle swarm optimization;SPSO-2011;CEC-2013;random topology;rotational invariance;benchmark testing;evolutionary computation;optimization},
+  doi = {10.1109/CEC.2013.6557848},
+}
+
 @Comment{jabref-meta: databaseType:bibtex;}

--- a/docs/source/refs.bib
+++ b/docs/source/refs.bib
@@ -906,19 +906,6 @@ url = {https://www.sciencedirect.com/science/article/pii/0010465575900399},
 author = {F. James and M. Roos}
 }
 
-@article{JAMES1975343,
-title = {Minuit - a system for function minimization and analysis of the parameter errors and correlations},
-journal = {Computer Physics Communications},
-volume = {10},
-number = {6},
-pages = {343-367},
-year = {1975},
-issn = {0010-4655},
-doi = {https://doi.org/10.1016/0010-4655(75)90039-9},
-url = {https://www.sciencedirect.com/science/article/pii/0010465575900399},
-author = {F. James and M. Roos}
-}
-
 @InProceedings{Kennedy1995,
   author={Kennedy, J. and Eberhart, R.},
   booktitle={Proceedings of ICNN'95 - International Conference on Neural Networks}, 

--- a/environment.yml
+++ b/environment.yml
@@ -7,6 +7,7 @@ dependencies:
   - python=3.10  # dev
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
+  - nevergrad  # dev, tests
   - jupyterlab  # dev, docs
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs

--- a/environment.yml
+++ b/environment.yml
@@ -7,7 +7,6 @@ dependencies:
   - python=3.10  # dev
   - cyipopt>=1.4.0  # dev, tests
   - pygmo>=2.19.0  # dev, tests, docs
-  - nevergrad  # dev, tests
   - jupyterlab  # dev, docs
   - nlopt  # dev, tests, docs
   - pip  # dev, tests, docs
@@ -40,6 +39,7 @@ dependencies:
   - annotated-types  # dev, tests
   - iminuit  # dev, tests
   - pip:  # dev, tests, docs
+      - nevergrad  # dev, tests
       - DFO-LS>=1.5.3  # dev, tests
       - Py-BOBYQA  # dev, tests
       - fides==0.7.4  # dev, tests

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -380,5 +380,6 @@ module = [
     "annotated_types",
     "pdbp",
     "iminuit",
+    "nevergrad",
   ]
 ignore_missing_imports = true

--- a/src/optimagic/algorithms.py
+++ b/src/optimagic/algorithms.py
@@ -18,6 +18,7 @@ from optimagic.optimizers.iminuit_migrad import IminuitMigrad
 from optimagic.optimizers.ipopt import Ipopt
 from optimagic.optimizers.nag_optimizers import NagDFOLS, NagPyBOBYQA
 from optimagic.optimizers.neldermead import NelderMeadParallel
+from optimagic.optimizers.nevergrad_optimizers import NevergradPSO
 from optimagic.optimizers.nlopt_optimizers import (
     NloptBOBYQA,
     NloptCCSAQ,
@@ -171,6 +172,7 @@ class BoundedGlobalGradientFreeNonlinearConstrainedParallelAlgorithms(AlgoSelect
 
 @dataclass(frozen=True)
 class BoundedGlobalGradientFreeParallelScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -364,6 +366,7 @@ class BoundedGlobalGradientFreeNonlinearConstrainedAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGlobalGradientFreeScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
     nlopt_direct: Type[NloptDirect] = NloptDirect
     nlopt_esch: Type[NloptESCH] = NloptESCH
@@ -403,6 +406,7 @@ class BoundedGlobalGradientFreeScalarAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGlobalGradientFreeParallelAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -458,6 +462,7 @@ class GlobalGradientFreeNonlinearConstrainedParallelAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class GlobalGradientFreeParallelScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -605,6 +610,7 @@ class BoundedGradientFreeNonlinearConstrainedParallelAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGradientFreeParallelScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -699,6 +705,7 @@ class BoundedGlobalNonlinearConstrainedParallelAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGlobalParallelScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -1025,6 +1032,7 @@ class GradientBasedNonlinearConstrainedScalarAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGlobalGradientFreeAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
     nlopt_direct: Type[NloptDirect] = NloptDirect
     nlopt_esch: Type[NloptESCH] = NloptESCH
@@ -1088,6 +1096,7 @@ class GlobalGradientFreeNonlinearConstrainedAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class GlobalGradientFreeScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
     nlopt_direct: Type[NloptDirect] = NloptDirect
     nlopt_esch: Type[NloptESCH] = NloptESCH
@@ -1131,6 +1140,7 @@ class GlobalGradientFreeScalarAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class GlobalGradientFreeParallelAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -1296,6 +1306,7 @@ class BoundedGradientFreeNonlinearConstrainedAlgorithms(AlgoSelection):
 @dataclass(frozen=True)
 class BoundedGradientFreeScalarAlgorithms(AlgoSelection):
     nag_pybobyqa: Type[NagPyBOBYQA] = NagPyBOBYQA
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_bobyqa: Type[NloptBOBYQA] = NloptBOBYQA
     nlopt_cobyla: Type[NloptCOBYLA] = NloptCOBYLA
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
@@ -1367,6 +1378,7 @@ class BoundedGradientFreeLeastSquaresAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGradientFreeParallelAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pounders: Type[Pounders] = Pounders
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
@@ -1448,6 +1460,7 @@ class GradientFreeNonlinearConstrainedParallelAlgorithms(AlgoSelection):
 @dataclass(frozen=True)
 class GradientFreeParallelScalarAlgorithms(AlgoSelection):
     neldermead_parallel: Type[NelderMeadParallel] = NelderMeadParallel
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -1516,6 +1529,7 @@ class BoundedGlobalNonlinearConstrainedAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGlobalScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
     nlopt_direct: Type[NloptDirect] = NloptDirect
     nlopt_esch: Type[NloptESCH] = NloptESCH
@@ -1564,6 +1578,7 @@ class BoundedGlobalScalarAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGlobalParallelAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -1632,6 +1647,7 @@ class GlobalNonlinearConstrainedParallelAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class GlobalParallelScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -1866,6 +1882,7 @@ class BoundedNonlinearConstrainedParallelAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedParallelScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -2124,6 +2141,7 @@ class GradientBasedLikelihoodAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class GlobalGradientFreeAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
     nlopt_direct: Type[NloptDirect] = NloptDirect
     nlopt_esch: Type[NloptESCH] = NloptESCH
@@ -2211,6 +2229,7 @@ class GradientFreeLocalAlgorithms(AlgoSelection):
 class BoundedGradientFreeAlgorithms(AlgoSelection):
     nag_dfols: Type[NagDFOLS] = NagDFOLS
     nag_pybobyqa: Type[NagPyBOBYQA] = NagPyBOBYQA
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_bobyqa: Type[NloptBOBYQA] = NloptBOBYQA
     nlopt_cobyla: Type[NloptCOBYLA] = NloptCOBYLA
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
@@ -2307,6 +2326,7 @@ class GradientFreeNonlinearConstrainedAlgorithms(AlgoSelection):
 class GradientFreeScalarAlgorithms(AlgoSelection):
     nag_pybobyqa: Type[NagPyBOBYQA] = NagPyBOBYQA
     neldermead_parallel: Type[NelderMeadParallel] = NelderMeadParallel
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_bobyqa: Type[NloptBOBYQA] = NloptBOBYQA
     nlopt_cobyla: Type[NloptCOBYLA] = NloptCOBYLA
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
@@ -2387,6 +2407,7 @@ class GradientFreeLeastSquaresAlgorithms(AlgoSelection):
 @dataclass(frozen=True)
 class GradientFreeParallelAlgorithms(AlgoSelection):
     neldermead_parallel: Type[NelderMeadParallel] = NelderMeadParallel
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pounders: Type[Pounders] = Pounders
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
@@ -2426,6 +2447,7 @@ class GradientFreeParallelAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedGlobalAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
     nlopt_direct: Type[NloptDirect] = NloptDirect
     nlopt_esch: Type[NloptESCH] = NloptESCH
@@ -2507,6 +2529,7 @@ class GlobalNonlinearConstrainedAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class GlobalScalarAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
     nlopt_direct: Type[NloptDirect] = NloptDirect
     nlopt_esch: Type[NloptESCH] = NloptESCH
@@ -2559,6 +2582,7 @@ class GlobalScalarAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class GlobalParallelAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -2823,6 +2847,7 @@ class BoundedScalarAlgorithms(AlgoSelection):
     iminuit_migrad: Type[IminuitMigrad] = IminuitMigrad
     ipopt: Type[Ipopt] = Ipopt
     nag_pybobyqa: Type[NagPyBOBYQA] = NagPyBOBYQA
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_bobyqa: Type[NloptBOBYQA] = NloptBOBYQA
     nlopt_ccsaq: Type[NloptCCSAQ] = NloptCCSAQ
     nlopt_cobyla: Type[NloptCOBYLA] = NloptCOBYLA
@@ -2923,6 +2948,7 @@ class BoundedLeastSquaresAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class BoundedParallelAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pounders: Type[Pounders] = Pounders
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
@@ -3024,6 +3050,7 @@ class NonlinearConstrainedParallelAlgorithms(AlgoSelection):
 @dataclass(frozen=True)
 class ParallelScalarAlgorithms(AlgoSelection):
     neldermead_parallel: Type[NelderMeadParallel] = NelderMeadParallel
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
     scipy_brute: Type[ScipyBrute] = ScipyBrute
@@ -3131,6 +3158,7 @@ class GradientFreeAlgorithms(AlgoSelection):
     nag_dfols: Type[NagDFOLS] = NagDFOLS
     nag_pybobyqa: Type[NagPyBOBYQA] = NagPyBOBYQA
     neldermead_parallel: Type[NelderMeadParallel] = NelderMeadParallel
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_bobyqa: Type[NloptBOBYQA] = NloptBOBYQA
     nlopt_cobyla: Type[NloptCOBYLA] = NloptCOBYLA
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
@@ -3201,6 +3229,7 @@ class GradientFreeAlgorithms(AlgoSelection):
 
 @dataclass(frozen=True)
 class GlobalAlgorithms(AlgoSelection):
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_crs2_lm: Type[NloptCRS2LM] = NloptCRS2LM
     nlopt_direct: Type[NloptDirect] = NloptDirect
     nlopt_esch: Type[NloptESCH] = NloptESCH
@@ -3334,6 +3363,7 @@ class BoundedAlgorithms(AlgoSelection):
     ipopt: Type[Ipopt] = Ipopt
     nag_dfols: Type[NagDFOLS] = NagDFOLS
     nag_pybobyqa: Type[NagPyBOBYQA] = NagPyBOBYQA
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_bobyqa: Type[NloptBOBYQA] = NloptBOBYQA
     nlopt_ccsaq: Type[NloptCCSAQ] = NloptCCSAQ
     nlopt_cobyla: Type[NloptCOBYLA] = NloptCOBYLA
@@ -3470,6 +3500,7 @@ class ScalarAlgorithms(AlgoSelection):
     ipopt: Type[Ipopt] = Ipopt
     nag_pybobyqa: Type[NagPyBOBYQA] = NagPyBOBYQA
     neldermead_parallel: Type[NelderMeadParallel] = NelderMeadParallel
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_bobyqa: Type[NloptBOBYQA] = NloptBOBYQA
     nlopt_ccsaq: Type[NloptCCSAQ] = NloptCCSAQ
     nlopt_cobyla: Type[NloptCOBYLA] = NloptCOBYLA
@@ -3598,6 +3629,7 @@ class LikelihoodAlgorithms(AlgoSelection):
 @dataclass(frozen=True)
 class ParallelAlgorithms(AlgoSelection):
     neldermead_parallel: Type[NelderMeadParallel] = NelderMeadParallel
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     pounders: Type[Pounders] = Pounders
     pygmo_gaco: Type[PygmoGaco] = PygmoGaco
     pygmo_pso_gen: Type[PygmoPsoGen] = PygmoPsoGen
@@ -3646,6 +3678,7 @@ class Algorithms(AlgoSelection):
     nag_dfols: Type[NagDFOLS] = NagDFOLS
     nag_pybobyqa: Type[NagPyBOBYQA] = NagPyBOBYQA
     neldermead_parallel: Type[NelderMeadParallel] = NelderMeadParallel
+    nevergrad_pso: Type[NevergradPSO] = NevergradPSO
     nlopt_bobyqa: Type[NloptBOBYQA] = NloptBOBYQA
     nlopt_ccsaq: Type[NloptCCSAQ] = NloptCCSAQ
     nlopt_cobyla: Type[NloptCOBYLA] = NloptCOBYLA

--- a/src/optimagic/config.py
+++ b/src/optimagic/config.py
@@ -100,6 +100,14 @@ else:
     IS_IMINUIT_INSTALLED = True
 
 
+try:
+    import nevergrad  # noqa: F401
+except ImportError:
+    IS_NEVERGRAD_INSTALLED = False
+else:
+    IS_NEVERGRAD_INSTALLED = True
+
+
 # ======================================================================================
 # Check if pandas version is newer or equal to version 2.1.0
 # ======================================================================================

--- a/src/optimagic/optimizers/nevergrad_optimizers.py
+++ b/src/optimagic/optimizers/nevergrad_optimizers.py
@@ -39,6 +39,7 @@ class NevergradPSO(Algorithm):
     transform: Literal["arctan", "gaussian", "identity"] = "arctan"
     population_size: int | None = None
     n_cores: int = 1
+    seed: int | None = None
     stopping_maxfun: PositiveInt = STOPPING_MAXFUN_GLOBAL
     inertia: float = 0.5 / math.log(2.0)
     cognitive: float = 0.5 + math.log(2.0)
@@ -63,6 +64,9 @@ class NevergradPSO(Algorithm):
                 shape=x0.shape, lower=problem.bounds.lower, upper=problem.bounds.upper
             )
         )
+
+        if self.seed is not None:
+            instrum.random_state.seed(self.seed)
 
         optimizer = ng.optimizers.ConfPSO(
             transform=self.transform,

--- a/src/optimagic/optimizers/nevergrad_optimizers.py
+++ b/src/optimagic/optimizers/nevergrad_optimizers.py
@@ -1,0 +1,91 @@
+"""Implement nevergrad optimizers."""
+
+import math
+from dataclasses import dataclass
+from typing import Literal
+
+import numpy as np
+from numpy.typing import NDArray
+
+from optimagic import mark
+from optimagic.config import IS_NEVERGRAD_INSTALLED
+from optimagic.exceptions import NotInstalledError
+from optimagic.optimization.algo_options import STOPPING_MAXFUN_GLOBAL
+from optimagic.optimization.algorithm import Algorithm, InternalOptimizeResult
+from optimagic.optimization.internal_optimization_problem import (
+    InternalOptimizationProblem,
+)
+from optimagic.typing import AggregationLevel, PositiveInt
+
+if IS_NEVERGRAD_INSTALLED:
+    import nevergrad as ng
+
+
+@mark.minimizer(
+    name="nevergrad_pso",
+    solver_type=AggregationLevel.SCALAR,
+    is_available=IS_NEVERGRAD_INSTALLED,
+    is_global=True,
+    needs_jac=False,
+    needs_hess=False,
+    supports_parallelism=True,
+    supports_bounds=True,
+    supports_linear_constraints=False,
+    supports_nonlinear_constraints=False,
+    disable_history=True,
+)
+@dataclass(frozen=True)
+class NevergradPSO(Algorithm):
+    transform: Literal["arctan", "gaussian", "identity"] = "arctan"
+    population_size: int | None = None
+    n_cores: int = 1
+    stopping_maxfun: PositiveInt = STOPPING_MAXFUN_GLOBAL
+    inertia: float = 0.5 / math.log(2.0)
+    cognitive: float = 0.5 + math.log(2.0)
+    social: float = 0.5 + math.log(2.0)
+    quasi_opp_init: bool = False
+    speed_quasi_opp_init: bool = False
+    special_speed_quasi_opp_init: bool = False
+
+    def _solve_internal_problem(
+        self, problem: InternalOptimizationProblem, x0: NDArray[np.float64]
+    ) -> InternalOptimizeResult:
+        if not IS_NEVERGRAD_INSTALLED:
+            raise NotInstalledError(
+                "The nevergrad_pso optimizer requires the 'nevergrad' package to be "
+                "installed. You can install it with `pip install nevergrad`. "
+                "Visit https://facebookresearch.github.io/nevergrad/getting_started.html"
+                " for more detailed installation instructions."
+            )
+
+        instrum = ng.p.Instrumentation(
+            ng.p.Array(
+                shape=x0.shape, lower=problem.bounds.lower, upper=problem.bounds.upper
+            )
+        )
+
+        optimizer = ng.optimizers.ConfPSO(
+            transform=self.transform,
+            popsize=self.population_size,
+            omega=self.inertia,
+            phip=self.cognitive,
+            phig=self.social,
+            qo=self.quasi_opp_init,
+            sqo=self.speed_quasi_opp_init,
+            so=self.special_speed_quasi_opp_init,
+        )(
+            parametrization=instrum,
+            budget=self.stopping_maxfun,
+            num_workers=self.n_cores,
+        )
+
+        recommendation = optimizer.minimize(problem.fun)
+
+        result = InternalOptimizeResult(
+            x=recommendation.value[0][0],
+            fun=recommendation.loss,
+            success=True,
+            n_fun_evals=optimizer.num_ask,
+        )
+
+        return result

--- a/src/optimagic/optimizers/nevergrad_optimizers.py
+++ b/src/optimagic/optimizers/nevergrad_optimizers.py
@@ -79,11 +79,11 @@ class NevergradPSO(Algorithm):
             num_workers=self.n_cores,
         )
 
-        while optimizer.num_ask < optimizer.budget:
+        while optimizer.num_ask < self.stopping_maxfun:
             x_list = [
                 optimizer.ask()
                 for _ in range(
-                    min(optimizer.num_workers, optimizer.budget - optimizer.num_ask)
+                    min(self.n_cores, self.stopping_maxfun - optimizer.num_ask)
                 )
             ]
             losses = problem.batch_fun(
@@ -99,6 +99,8 @@ class NevergradPSO(Algorithm):
             fun=recommendation.loss,
             success=True,
             n_fun_evals=optimizer.num_ask,
+            n_jac_evals=0,
+            n_hess_evals=0,
         )
 
         return result

--- a/src/optimagic/optimizers/nevergrad_optimizers.py
+++ b/src/optimagic/optimizers/nevergrad_optimizers.py
@@ -32,7 +32,7 @@ if IS_NEVERGRAD_INSTALLED:
     supports_bounds=True,
     supports_linear_constraints=False,
     supports_nonlinear_constraints=False,
-    disable_history=True,
+    disable_history=False,
 )
 @dataclass(frozen=True)
 class NevergradPSO(Algorithm):
@@ -79,7 +79,20 @@ class NevergradPSO(Algorithm):
             num_workers=self.n_cores,
         )
 
-        recommendation = optimizer.minimize(problem.fun)
+        while optimizer.num_ask < optimizer.budget:
+            x_list = [
+                optimizer.ask()
+                for _ in range(
+                    min(optimizer.num_workers, optimizer.budget - optimizer.num_ask)
+                )
+            ]
+            losses = problem.batch_fun(
+                [x.value[0][0] for x in x_list], n_cores=self.n_cores
+            )
+            for x, loss in zip(x_list, losses, strict=True):
+                optimizer.tell(x, loss)
+
+        recommendation = optimizer.provide_recommendation()
 
         result = InternalOptimizeResult(
             x=recommendation.value[0][0],

